donut/modules/voting/helpers.py

```diff
@@ -0,0 +1,395 @@
+        SELECT question_id
+        FROM survey_questions NATURAL JOIN survey_responses NATURAL JOIN users
+        WHERE survey_id = %s AND username = %s
+        LIMIT 1
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(responses_query, [survey['survey_id'], username])
+        if cursor.fetchone(): return 'Already completed'
+
+
+def restrict_edit_access(survey, allow_after_close):
+    if not survey:
+        return 'Invalid access key'
+    user_id = get_user_id(flask.session.get('username'))
+    if user_id != survey['creator']:
+        return 'You are not the creator of this survey'
+    if not allow_after_close and survey['end_time'] < datetime.now():
+        return 'Cannot modify a survey after it has closed'
+
+
+def set_responses(question_ids, responses):
+    user_id = get_user_id(flask.session['username'])
+    query = """
+        INSERT INTO survey_responses(question_id, user_id, response)
+        VALUES (%s, %s, %s)
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        for question_id, response in zip(question_ids, responses):
+            cursor.execute(query, [question_id, user_id, response])
+
+
+def some_responses_for_survey(survey_id):
+    query = """
+        SELECT user_id
+        FROM survey_responses NATURAL JOIN survey_questions
+        WHERE survey_id = %s
+        LIMIT 1
+    """
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(query, [survey_id])
+        return cursor.fetchone() is not None
+
+
+def get_results(survey_id):
+    questions_query = """
+        SELECT question_id, title, description, type_id AS type, list_order, choices
+        FROM survey_questions NATURAL JOIN survey_question_types
+        WHERE survey_id = %s
+        ORDER BY list_order
+    """
+    choices_query = """
+        SELECT choice_id, choice
+        FROM survey_question_choices
+        WHERE question_id = %s
+        ORDER BY choice_id
+    """
+    responses_query = """
+        SELECT response
+        FROM survey_questions NATURAL JOIN survey_responses
+        WHERE question_id = %s
+    """
+    name_query = 'SELECT full_name FROM members_full_name WHERE user_id = %s'
+    question_types = get_question_types()
+    count_types = set([
+        question_types['Dropdown'], question_types['Short text'],
+        question_types['Long text']
+    ])
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(questions_query, [survey_id])
+        questions = cursor.fetchall()
+        for question in questions:
+            if question['choices']:
+                cursor.execute(choices_query, [question['question_id']])
+                question['choices'] = {
+                    choice['choice_id']: choice['choice']
+                    for choice in cursor.fetchall()
+                }
+            else:
+                del question['choices']
+
+            def resolve_name(vote):
+                if vote is None: return NO
```

Comment:
seems doubtful that we would need to represent a "NO" vote as None, why can't it just be a "NO" string?

Reply:
The format of the elected positions responses is arrays of choice_ids and user_ids; I didn't want to allow arbitrary strings, so I just allowed null. I imagine it makes it slightly faster for the ranked pairs algorithm because it is faster to check None for equality than checking strings for equality.